### PR TITLE
Implemented a ClientFactory to obtain a new instance of a ChatGptClient

### PR DESCRIPTION
**Pull request by nicolaparo**
Adding the `AddChatGptClientFactory` to add the ClientFactory.

Remarks:
- I had to change the visibility of `ChatGptServiceCollectionExtensions.ConfigureHttpClient` from `private` to `internal` in order to be able to correctly configure the httpclient in the factory `CreateClient` method.
- I changed also `ChatGptOptions` from class to `record` in order to support the `defaultOptions` cloning using `with { }`. This way the changes to the options are applied to a separate copy of the object in the `CreateClient` method.

### CODE REVIEW

**Comment by nicolaparo** [comment]
This pr might solve Issue #45

File: src/ChatGptNet/ChatGptClientFactory.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    private readonly IMemoryCache memoryCache;
+    private readonly ChatGptOptions defaultOptions;
+
+    public ChatGptClientFactory(IMemoryCache memoryCache, ChatGptOptions defaultOptions)
+    {
+        this.memoryCache = memoryCache;
+        this.defaultOptions = defaultOptions;
+    }
+
+    public IChatGptClient CreateClient(Action<ChatGptOptions> setupAction)
```
Comment (marcominerva) [inline src/ChatGptNet/ChatGptClientFactory.cs:16]: The parameter should be nullable and optional, if I want to create a client with the default options and I don't need to change anything else. Moreover, an overload the accepts a `ServiceProvider` as argument is needed, to cover a scenario like this one: https://github.com/marcominerva/ChatGptNet/blob/2193d2c7b9656d337cbd24b344b8ea3bba995fbb/src/ChatGptNet/ChatGptServiceCollectionExtensions.cs#L72

It is necessary when I need to provide options that have external dependencies (for example, an Api Key that depends on the current user).

File: src/ChatGptNet/ChatGptServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -9,6 +9,52 @@ namespace ChatGptNet;
...
+    public static IServiceCollection AddChatGptClientFactory(this IServiceCollection services, Action<ChatGptOptions> setupAction)
+    {
+        ArgumentNullException.ThrowIfNull(services);
+        ArgumentNullException.ThrowIfNull(setupAction);
+
+        var options = new ChatGptOptions();
+        setupAction.Invoke(options);
+
+        services.AddMemoryCache();
+        services.AddSingleton<IChatGptClientFactory, ChatGptClientFactory>(
```
Comment (marcominerva) [inline src/ChatGptNet/ChatGptServiceCollectionExtensions.cs:27]: Why are you using this method and not simply, after adding `ChatGptOptions `itself to the service collection, like in https://github.com/marcominerva/ChatGptNet/blob/2193d2c7b9656d337cbd24b344b8ea3bba995fbb/src/ChatGptNet/ChatGptServiceCollectionExtensions.cs#L28):

`services.AddSingleton<IChatGptClientFactory, ChatGptClientFactory>();`

File: src/ChatGptNet/ChatGptServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -9,6 +9,52 @@ namespace ChatGptNet;
 /// </summary>
 public static class ChatGptServiceCollectionExtensions
 {
+    /// <summary>
+    /// Registers a <see cref="ChatGptClientFactory"/> instance with the specified options.
+    /// </summary>
+    /// <param name="services">The <see cref="IServiceCollection"/> to add services to.</param>
+    /// <param name="setupAction">The <see cref="Action{ChatGptOptions}"/> to configure the provided <see cref="ChatGptOptions"/>.</param>
+    /// <returns>A reference to this instance after the operation has completed.</returns>
+    public static IServiceCollection AddChatGptClientFactory(this IServiceCollection services, Action<ChatGptOptions> setupAction)
```
Comment (marcominerva) [inline src/ChatGptNet/ChatGptServiceCollectionExtensions.cs:18]: If I'm creating a factory, there is the case in which I don't want to directly provide an Action to setup `ChatGtpOptions`, so this parameter should be nullable and optional.

**Review by marcominerva — CHANGES_REQUESTED** [review]

File: src/ChatGptNet/ChatGptClientFactory.cs (hunk 1 above)
Comment (nicolaparo) [reply src/ChatGptNet/ChatGptClientFactory.cs]: Implemented the `CreateClient` method with 3 overloads: a parameterless version that uses the default options, plus two overloads that accept an `Action<ChatGptOptions>` or `Action<IServiceProvider, ChatGptOptions>`

File: src/ChatGptNet/ChatGptServiceCollectionExtensions.cs (hunk 3 above)
Comment (nicolaparo) [reply src/ChatGptNet/ChatGptServiceCollectionExtensions.cs]: Added now an `AddChatGptClientFactory(this IServiceCollection services)` that allows to customize later the client

**Comment by marcominerva** [comment]
@nicolaparo, sorry for the long delay. Do you mind updating your PR to the latest commit of the library and then requesting a review? Thanks

**Pull request merged**